File: src/main/java/com/github/dkoval/leetcode/challenge/PermutationInString.java
package com.github.dkoval.leetcode.challenge;

/**
 * <a href="https://leetcode.com/problems/permutation-in-string/">Permutation in String</a>
 * <p>
 * Given two strings s1 and s2, return true if s2 contains a
 * permutation
 * of s1, or false otherwise.
 * <p>
 * In other words, return true if one of s1's permutations is the substring of s2.
 * <p>
 * Constraints:
 * <ul>
 *  <li>1 <= s1.length, s2.length <= 10^4</li>
 *  <li>s1 and s2 consist of lowercase English letters</li>
 * </ul>
 */
public interface PermutationInString {

    boolean checkInclusion(String s1, String s2);

    class PermutationInStringRev2 implements PermutationInString {

        @Override
        public boolean checkInclusion(String s1, String s2) {
            int n1 = s1.length();
            int n2 = s2.length();

            if (n2 < n1) {
                return false;
            }

            // idea: sliding window
            int[] counts1 = new int[26];
            int[] counts2 = new int[26];
            for (int i = 0; i < n2; i++) {
                // frequency table of a sliding window of size n1
                counts2[s2.charAt(i) - 'a']++;
                if (i >= n1) {
                    // Example: n1 = 2
                    // 0, 1, 2, 3
                    // ----
                    //    ----
                    //       ----
                    // remove the 1st symbol of the previous window
                    counts2[s2.charAt(i - n1) - 'a']--;
                }

                // frequency table of s1
                if (i < n1) {
                    counts1[s1.charAt(i) - 'a']++;
                }

                if (i >= n1 - 1 && matches(counts1, counts2)) {
                    return true;
                }
            }
            return false;
        }

        private boolean matches(int[] counts1, int[] counts2) {
            for (int i = 0; i < 26; i++) {
                if (counts1[i] != counts2[i]) {
                    return false;
                }
            }
            return true;
        }
    }

    class PermutationInStringRev3 implements PermutationInString {

        @Override
        public boolean checkInclusion(String s1, String s2) {
            int n1 = s1.length();
            int n2 = s2.length();

            if (n2 < n1) {
                return false;
            }

            // idea: sliding window of size n1
            int[] counts1 = new int[26];
            int[] counts2 = new int[26];
            for (int i = 0; i < n1; i++) {
                counts1[s1.charAt(i) - 'a']++;
                counts2[s2.charAt(i) - 'a']++;
            }

            // index i is the ending index of the current window
            for (int i = n1 - 1; i < n2; i++) {
                if (i > n1 - 1) {
                    // append the current symbol to the current window
                    counts2[s2.charAt(i) - 'a']++;
                    // remove the 1st symbol of the previous window
                    counts2[s2.charAt(i - n1) - 'a']--;
                }

                if (matches(counts1, counts2)) {
                    return true;
                }
            }
            return false;
        }

        private boolean matches(int[] counts1, int[] counts2) {
            for (int i = 0; i < 26; i++) {
                if (counts1[i] != counts2[i]) {
                    return false;
                }
            }
            return true;
        }
    }
}

File: src/main/kotlin/com/github/dkoval/leetcode/challenge/PermutationInString.kt
package com.github.dkoval.leetcode.challenge

object PermutationInStringRev1 : PermutationInString {

    private const val ALPHABET_SIZE = 26

    // O(l1 + 26 * (l2 - l1)) time | O(1) space - only counts1[] and counts2[] of size 26 are used
    override fun checkInclusion(s1: String, s2: String): Boolean {
        val l1 = s1.length
        val l2 = s2.length

        if (l1 > l2) {
            return false
        }

        // frequency table of string s1
        val counts1 = IntArray(ALPHABET_SIZE)
        // frequency table of a window of length l1 in string s2
        val counts2 = IntArray(ALPHABET_SIZE)
        for (i in 0 until l1) {
            counts1[s1[i] - 'a']++
            counts2[s2[i] - 'a']++
        }

        // sliding window: consider all windows of length l1 in string s2
        for (start in 0..(l2 - l1)) {
            if (start > 0) {
                // remove starting character of the previous windows from the frequency table of string s2
                counts2[s2[start - 1] - 'a']--
                // add ending character of the current window to the frequency table of string s2
                counts2[s2[start + l1 - 1] - 'a']++
            }
            if (matches(counts1, counts2)) {
                return true
            }
        }
        return false
    }

    private fun matches(counts1: IntArray, counts2: IntArray): Boolean {
        for (i in 0 until ALPHABET_SIZE) {
            if (counts1[i] != counts2[i]) {
                return false
            }
        }
        return true
    }
}